go/pkg/cas/upload.go

```diff
@@ -195,21 +224,49 @@ func (u *uploader) startProcessing(ctx context.Context, in *UploadInput) error {
 			return errors.WithStack(err)
 		}
 
-		_, err = u.visitFile(ctx, absPath, info)
+		_, err = u.visitPath(ctx, absPath, info, in.PathExclude)
 		return err
 	})
 	return nil
 }
 
-// visitFile visits the file/dir depending on its type (regular, dir, symlink).
+// visitPath visits the file/dir depending on its type (regular, dir, symlink).
 // Visits each file only once.
-func (u *uploader) visitFile(ctx context.Context, absPath string, info os.FileInfo) (dirEntry proto.Message, err error) {
-	node, err := u.fsCache.LoadOrStore(absPath, func() (interface{}, error) {
+//
+// If the file should be skipped, then returns (nil, nil).
+func (u *uploader) visitPath(ctx context.Context, absPath string, info os.FileInfo, pathExclude *regexp.Regexp) (dirEntry proto.Message, err error) {
+	// First, check if the file passes all filters.
+	if pathExclude != nil && pathExclude.MatchString(absPath) {
+		return nil, nil
+	}
+	// Call the callback only after checking the pathExclude.
+	if u.Callback != nil {
+		switch err := u.Callback(absPath, info.Mode()); {
+		case errors.Is(err, ErrSkip):
+			return nil, nil
+		case err != nil:
+			return nil, err
```

Comment:
I think you can remove the TODO on L432 original file?

Reply:
done